internal/pkg/sylog/sylog_test.go

```diff
@@ -0,0 +1,288 @@
+			expectedResult: os.Stderr,
+		},
+		{
+			name:           "valid logger",
+			loggerLevel:    1,
+			expectedResult: os.Stderr,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			SetLevel(tt.loggerLevel)
+			w := Writer()
+			if w != tt.expectedResult {
+				if w == ioutil.Discard {
+					fmt.Printf("%s returned ioutil.Discard\n", tt.name)
+				}
+				if w == os.Stderr {
+					fmt.Printf("%s returned os.Stderr\n", tt.name)
+				}
+				t.Fatal("Writer() did not return the expected io.Writer")
+			}
+		})
+	}
+}
+
+func TestWritef(t *testing.T) {
+	const str = "just a test"
+
+	tests := []struct {
+		name string
+		lvl  messageLevel
+	}{
+		{
+			name: "info",
+			lvl:  info,
+		},
+		{
+			name: "error",
+			lvl:  error,
+		},
+		{
+			name: "warning",
+			lvl:  warn,
+		},
+		{
+			name: "fatal",
+			lvl:  fatal,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var buf bytes.Buffer
+
+			SetLevel(int(tt.lvl))
+			DisableColor()
+
+			writef(&buf, tt.lvl, "%s", str)
+			expectedResult := prefix(tt.lvl) + str + "\n"
+			if buf.String() != expectedResult {
+				t.Fatalf("test %s returned %s instead of %s", tt.name, buf.String(), expectedResult)
+			}
+		})
+	}
+
+	// corner case
+	SetLevel(int(fatal))
+	expectedResult := ""
+	var buf bytes.Buffer
+	writef(&buf, info, "%s", str)
+	if buf.String() != expectedResult {
+		t.Fatalf("test returned %s instead of an empty string", buf.String())
+	}
+}
+
+func TestGetLevel(t *testing.T) {
+	tests := []struct {
+		name           string
+		lvl            messageLevel
+		expectedResult int
+	}{
+		{
+			name:           "fatal",
+			lvl:            fatal,
+			expectedResult: -4,
+		},
+		{
+			name:           "error",
+			lvl:            error,
+			expectedResult: -3,
+		},
+		{
+			name:           "warn",
+			lvl:            warn,
+			expectedResult: -2,
+		},
+		{
+			name:           "info",
+			lvl:            info,
+			expectedResult: 1,
+		},
+		{
+			name:           "verbose",
+			lvl:            verbose,
+			expectedResult: 2,
+		},
+		{
+			name:           "verbose2",
+			lvl:            verbose2,
+			expectedResult: 3,
+		},
+		{
+			name:           "verbose3",
+			lvl:            verbose3,
+			expectedResult: 4,
+		},
+		{
+			name:           "debug",
+			lvl:            debug,
+			expectedResult: 5,
+		},
+		{
+			name:           "invalid",
+			lvl:            messageLevel(-10),
+			expectedResult: -4,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			SetLevel(int(tt.lvl))
+			lvl := GetLevel()
+			if lvl != int(tt.lvl) {
+				t.Fatalf("test %s was expected to return %d but returned %d instead", tt.name, tt.expectedResult, lvl)
+			}
+		})
+	}
+}
+
+func TestGetenv(t *testing.T) {
+	str := GetEnvVar()
+	expectedResult := "SINGULARITY_MESSAGELEVEL="
+	if str[:25] != expectedResult {
+		t.Fatalf("Test returned %s instead of %s", str[:25], expectedResult)
+	}
+}
+
+func TestStderrOutput(t *testing.T) {
```

Comment:
Would it be possible to replace os.Stderr with an interface to a buffer or writer and capture the output for testing? 

I think you can use a pipe as well. Here is an example I found: https://play.golang.org/p/fXpK0ZhXXf

Reply:
Great idea and done.